### PR TITLE
[#1357] improvement(common): add `toSequence` support for configuration builder

### LINKED ISSUE #1357: [Improvement] Support sequence in the configuration
### What would you like to be improved?

Spark configuration have implemented to `toSeq` method for configuration options. It's necessary for Gravitino.

### How should we improve?

You can refer to https://github.com/apache/spark/blob/e8dfcd3081abe16b2115bb2944a2b1cb547eca8e/core/src/main/scala/org/apache/spark/internal/config/ConfigBuilder.scala#L126

### PR DESCRIPTION
**Pull request by austin362667**
### What changes were proposed in this pull request?

Add `toSequence()` support while dealing with a list of values in one conf entry.

### Why are the changes needed?

Fix #1357 

Just like Spark [does](https://github.com/apache/spark/blob/e8dfcd3081abe16b2115bb2944a2b1cb547eca8e/core/src/main/scala/org/apache/spark/internal/config/ConfigBuilder.scala#L126) and can be indeed useful for Gravitino.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Run `./gradlew test -PskipITs` should pass all.

### CODE REVIEW

**Comment by qqqttt123** [comment]
#1357 is an improvement. So we should use `improvement` instead of `fix`.

File: common/src/test/java/com/datastrato/gravitino/config/TestConfigEntry.java [hunk 1]
```diff
@@ -24,13 +25,60 @@ public void initializeConfigMap() {
...
+    List<Integer> valueList = testConf.readFrom(configMap);
+    Assertions.assertEquals(1, valueList.get(0));
+    Assertions.assertEquals(2, valueList.get(1));
+    Assertions.assertEquals(3, valueList.get(2));
+
+    valueList.set(0, 4);
+    valueList.set(1, 5);
+    valueList.set(2, 6);
+    testConf.writeTo(configMap, valueList);
+    Assertions.assertEquals("4,5,6", configMap.get("gravitino.test.integerList"));
```
Comment (qqqttt123) [inline common/src/test/java/com/datastrato/gravitino/config/TestConfigEntry.java:79]: Could you add some cases with exception? 
For example, "4,xxx,6"

File: common/src/test/java/com/datastrato/gravitino/config/TestConfigEntry.java [hunk 2]
```diff
@@ -24,13 +25,60 @@ public void initializeConfigMap() {
...
+    ConfigEntry<List<Integer>> testConf =
+        new ConfigBuilder("gravitino.test.integerList")
+            .doc("test")
+            .internal()
+            .intConf()
+            .toSequence()
+            .create();
+
+    List<Integer> valueList = testConf.readFrom(configMap);
+    Assertions.assertEquals(1, valueList.get(0));
```
Comment (qqqttt123) [inline common/src/test/java/com/datastrato/gravitino/config/TestConfigEntry.java:71]: Could this method work with `checkValue`?
For example,
```
        new ConfigBuilder("gravitino.test.integerList")
            .doc("test")
            .internal()
            .intConf()
            .checkValue(xxxx)
            .toSequence()
            .create();
```

**Comment by qqqttt123** [comment]
> ### What changes were proposed in this pull request?
> Add `toSequence()` support while dealing with a list of values in one conf_entry (strings, or integers).
> 
> > I am new to Java, and I am wondering if you could give me some suggestions on this PR. I have implemented the `toSequence()` API for arbitrary type_converter.
> 
> However, I ran into one issue and three concerns.
> 
> 1. I haven't figured out how to tackle the `defaultValue` if it's a value list.
>    e.g.,
> 
> ```java
>  new ConfigBuilder("gravitino.test.stringList")
>       .doc("test")
>       .internal()
>       .intConf()
>       .toSequence()
>       // .create(); (worked! can pass testing)
>       .createWithDefault("test-string0,test-string1,test-string2"); // (obviously not working.. can not pass testing)
> ```
> 
> 2. How about `str2seq` delimiter? Should we let the user set it by themselves?
> 3. Should we let users set more than two kinds of type within one entry? It might be difficult to handle.
>    e.g., entries like `{key: ['text', 0, 2, 4, 'str', 5, '']` mixes up with integers and strings.
> 4. Should we wrap these two util functions `strToSeq()`, and `seqToStr()` into another util file for modulization? See [here1](https://github.com/apache/spark/blob/e8dfcd3081abe16b2115bb2944a2b1cb547eca8e/core/src/main/scala/org/apache/spark/internal/config/ConfigBuilder.scala#L49C11-L49C22), [here2](https://github.com/apache/spark/blob/e8dfcd3081abe16b2115bb2944a2b1cb547eca8e/core/src/main/scala/org/apache/spark/util/Utils.scala#L2693)
> 
> ### Why are the changes needed?
> Fix #1357
> 
> Just like Spark [does](https://github.com/apache/spark/blob/e8dfcd3081abe16b2115bb2944a2b1cb547eca8e/core/src/main/scala/org/apache/spark/internal/config/ConfigBuilder.scala#L126) and can be indeed useful for Gravitino.
> 
> ### Does this PR introduce _any_ user-facing change?
> No
> 
> ### How was this patch tested?
> Run `./gradlew test -PskipITs` should pass all.

1. You can refer to the implement of Spark how to handle the default value.
2. It's ok to use `,`. `,` seems to be used most widely.
3. If we mixed multple types, we may throw exception.
4. Maybe apache.common.lang3 `StringUtils` has already similar methods.

File: common/src/test/java/com/datastrato/gravitino/config/TestConfigEntry.java (hunk 1 above)
Comment (austin362667) [reply common/src/test/java/com/datastrato/gravitino/config/TestConfigEntry.java:104]: sure, done

File: common/src/test/java/com/datastrato/gravitino/config/TestConfigEntry.java (hunk 2 above)
Comment (austin362667) [reply common/src/test/java/com/datastrato/gravitino/config/TestConfigEntry.java:96]: I think it can and it does after my test.
But, `checkValue()`, should only work after function `toSequence()` is invoked as expected.

**Comment by austin362667** [comment]
> You can refer to the implement of Spark how to handle the default value.
It's ok to use ,. , seems to be used most widely.
If we mixed multple types, we may throw exception.
Maybe apache.common.lang3 StringUtils has already similar methods.

Thank you so much for replying!

File: common/src/test/java/com/datastrato/gravitino/config/TestConfigEntryList.java [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+  }
+
+  @Test
+  public void testConfWithDefaultValue() {
+    ConfigEntry<List<String>> testConf1 =
+        new ConfigBuilder("gravitino.test.list")
+            .doc("test")
+            .internal()
+            .stringConf()
+            .checkValue(value -> value == "test-string-1,test-string-2,test-string-3", "error")
```
Comment (qqqttt123) [inline common/src/test/java/com/datastrato/gravitino/config/TestConfigEntryList.java:40]: Maybe we should call method `readFrom` for `testConf1`. Because the method `checkValue` is called when we call the method `readFrom`.

Comment (austin362667) [reply common/src/test/java/com/datastrato/gravitino/config/TestConfigEntryList.java]: I see, sure! thx for heads-up~

File: common/src/main/java/com/datastrato/gravitino/config/ConfigEntry.java [hunk 4]
```diff
@@ -25,6 +27,7 @@ public class ConfigEntry<T> {
   @Getter private List<String> alternatives;
 
   @Getter private T defaultValue;
+  @Getter private List<T> defaultValueList;
```
Comment (qqqttt123) [inline common/src/main/java/com/datastrato/gravitino/config/ConfigEntry.java:30]: Maybe we don't need this field.

Comment (austin362667) [reply common/src/main/java/com/datastrato/gravitino/config/ConfigEntry.java]: You are right.  We don't need this.

File: common/src/main/java/com/datastrato/gravitino/config/ConfigEntry.java [hunk 5]
```diff
@@ -220,7 +262,6 @@ public T readFrom(Map<String, String> properties) throws NoSuchElementException
         throw new NoSuchElementException("No configuration found for key " + key);
       }
     }
-
```
Comment (qqqttt123) [inline common/src/main/java/com/datastrato/gravitino/config/ConfigEntry.java:223]: Unnecessary change?

**Review by qqqttt123 — APPROVED** [review]
LGTM, just one nit.

File: common/src/main/java/com/datastrato/gravitino/config/ConfigEntry.java (hunk 5 above)
Comment (austin362667) [reply common/src/main/java/com/datastrato/gravitino/config/ConfigEntry.java]: Oops... For sure!

**Review by qqqttt123 — APPROVED** [review]

**Comment by qqqttt123** [comment]
@austin362667 Could you update your description of pull request?

**Comment by austin362667** [comment]
@qqqttt123  by description you mean like the title?

**Comment by qqqttt123** [comment]
> @qqqttt123 by description you mean like the title?

Now, it's ok. The description contains some issues which we have solved. We should remove those issues. If no one has another suggestion until tomorrow, I will merge this pr.

**Comment by qqqttt123** [comment]
Thanks @austin362667 for your contribution, merged to master.
